pocs/sensors/arduino_io.py

```diff
@@ -0,0 +1,209 @@
+        ports = list_arduino_ports()
+    if not logger:
+        logger = get_root_logger()
+    result = []
+    for port in ports:
+        board_name = detect_board_on_port(port, logger)
+        if board_name:
+            result.append((board_name, port))
+    return result
+
+
+# Note: list_arduino_ports is modified by test_arduino_io.py, so if changing
+# this import, the test will also need to be updated.
+def list_arduino_ports():
+    """Find devices (paths or URLs) that appear to be Arduinos.
+
+    Returns:
+        a list of strings; device paths (e.g. /dev/ttyACM1) or URLs (e.g. rfc2217://host:port
+        arduino_simulator://?board=camera).
+    """
+    ports = rs232.get_serial_port_info()
+    return [p.device for p in ports
+            if 'arduino' in p.description.lower() or 'arduino' in p.manufacturer.lower()]
+
+
+def detect_board_on_port(port, logger=None):
+    """Open a port and determine which type of board its producing output.
+
+    Returns: board_name if we can read a line of JSON from the
+        port, parse it and find a 'name' attribute in the top-level object.
+        Else returns None.
+    """
+    if not logger:
+        logger = get_root_logger()
+    logger.debug('Attempting to connect to serial port: {}'.format(port))
+    serial_reader = None
+    try:
+        # First open a connection to the device.
+        try:
+            serial_reader = open_serial_device(port)
+            if not serial_reader.is_connected:
+                serial_reader.connect()
+            logger.debug('Connected to {}', port)
+        except Exception as e:
+            logger.warning('Could not connect to port: {}'.format(port))
+            return None
+        try:
+            reading = serial_reader.get_and_parse_reading(retry_limit=3)
+            if not reading:
+                return None
+            (ts, data) = reading
+            if isinstance(data, dict) and 'name' in data and isinstance(data['name'], str):
+                return data['name']
+            logger.warning('Unable to find board name in reading: {!r}', reading)
+            return None
+        except Exception as e:
+            logger.error('Exception while auto-detecting port {!r}: {!r}'.format(port, e))
+    finally:
+        if serial_reader:
+            serial_reader.disconnect()
+
+
+def open_serial_device(port, serial_config=None, **kwargs):
+    # Using a long timeout (2 times the report interval) rather than retries which can just break
+    # a JSON line into two unparseable fragments.
+    sd_kwargs = dict(baudrate=9600, retry_limit=1, retry_delay=0, timeout=4.0, name=port)
+    if serial_config:
+        sd_kwargs.update(serial_config)
+    sd_kwargs.update(kwargs)
+    sd_kwargs['port'] = port
+    return rs232.SerialData(**sd_kwargs)
+
+
+class ArduinoIO(object):
+    """Reads the output from an arduino, and exposes the relays for change."""
+
+    def __init__(self, board_name, port, output_queue, serial_config=None):
```

Comment:
What is the overall picture of how this will be used? I mostly ask because I think it's a bit odd to pass in a `queue` rather than to have it create a queue.

Reply:
The aim is for it to deliver readings to a thread that will take care of writing to the database (or wherever). The advantage of providing the queue to the object is that we can use the same queue for multiple Arduinos, and then the db writer thread only needs to read from a single queue.